### ISSUE TITLE
Error when linting code with generic arrays

### ISSUE THREAD

**Issue by nbgraham** [golangci/golangci-lint#4332]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

Maybe this is just bad code, but I imagine it shouldn't throw an error when linting?

Seems to switch randomly between these two errors
```
ERRO Running error: 1 error occurred:
        * can't run linter goanalysis_metalinter: goanalysis_metalinter: nilness: package "main" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled type *ir.SliceToArray
```

```
ERRO Running error: 1 error occurred:
        * can't run linter goanalysis_metalinter: goanalysis_metalinter: SA4023: package "main" (isInitialPkg: true, needAnalyzeSource: true): interface conversion: interface {} is nil, not *nilness.Result
```

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version v1.55.2 built with go1.21.3 from (unknown, mod sum: "h1:yllEIsSJ7MtlDBwDJ9IMBkyEUz2fYE0b5B8IUgO1oP8=") on (unknown)
```

</details>

### Configuration

<details>

None
```console
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.21.3 darwin/arm64
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/X/Library/Caches/go-build'
GOENV='/Users/X/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/X/go/pkg/mod'
GOOS='darwin'
GOPATH='/Users/X/go'
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.21.3/libexec'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.21.3/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.21.3'
GCCGO='gccgo'
AR='ar'
CC='cc'
CXX='c++'
CGO_ENABLED='1'
GOMOD='/Users/X/Developer/golangci-lint-slice-to-array/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/6t/_0d1whkd0c3_qh6jxwrv9h4h0000gq/T/go-build2612010489=/tmp/go-build -gno-record-gcc-switches -fno-common
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/PYY388/Developer/golangci-lint-slice-to-array /Users/PYY388/Developer /Users/PYY388 /Users /] 
INFO [lintersdb] Active 6 linters: [errcheck gosimple govet ineffassign staticcheck unused] 
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|files|imports|deps|exports_file|name) took 105.071875ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 173.708µs 
ERRO [linters_context/goanalysis] nilness: panic during analysis: internal error: unhandled type *ir.SliceToArray, goroutine 826 [running]:
runtime/debug.Stack()
        /opt/homebrew/Cellar/go/1.21.3/libexec/src/runtime/debug/stack.go:24 +0x64
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:105 +0x4c
panic({0x105237540?, 0x14004e1fd90?})
        /opt/homebrew/Cellar/go/1.21.3/libexec/src/runtime/panic.go:914 +0x218
honnef.co/go/tools/analysis/facts/nilness.impl.func1({0x105443da0, 0x140020afdc0})
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:239 +0x804
honnef.co/go/tools/analysis/facts/nilness.impl.func1({0x1054432a8, 0x14001bd5800})
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:147 +0x938
honnef.co/go/tools/analysis/facts/nilness.impl(0x14001ce2750, 0x140020adb80, 0x140013cbc88)
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:246 +0x2c8
honnef.co/go/tools/analysis/facts/nilness.run(0x14001ce2750)
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:66 +0xf4
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0x140014958a0)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:195 +0x900
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:113 +0x20
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0x14001328280, {0x104efef4a, 0x7}, 0x140010f7730)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0x0?)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:112 +0x78
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0x140014958a0)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb0
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze in goroutine 102
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x184 
INFO [linters_context/goanalysis] analyzers took 1.240967294s with top 10 stages: buildir: 511.027791ms, fact_purity: 123.238207ms, nilness: 106.459876ms, SA5012: 99.956416ms, ctrlflow: 96.851375ms, typedness: 95.723543ms, printf: 94.33425ms, fact_deprecated: 92.118459ms, inspect: 19.33271ms, SA4006: 239.333µs 
ERRO [runner] Panic: nilness: package "main" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled type *ir.SliceToArray: goroutine 826 [running]:
runtime/debug.Stack()
        /opt/homebrew/Cellar/go/1.21.3/libexec/src/runtime/debug/stack.go:24 +0x64
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:109 +0x224
panic({0x105237540?, 0x14004e1fd90?})
        /opt/homebrew/Cellar/go/1.21.3/libexec/src/runtime/panic.go:914 +0x218
honnef.co/go/tools/analysis/facts/nilness.impl.func1({0x105443da0, 0x140020afdc0})
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:239 +0x804
honnef.co/go/tools/analysis/facts/nilness.impl.func1({0x1054432a8, 0x14001bd5800})
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:147 +0x938
honnef.co/go/tools/analysis/facts/nilness.impl(0x14001ce2750, 0x140020adb80, 0x140013cbc88)
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:246 +0x2c8
honnef.co/go/tools/analysis/facts/nilness.run(0x14001ce2750)
        /Users/PYY388/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:66 +0xf4
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0x140014958a0)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:195 +0x900
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:113 +0x20
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0x14001328280, {0x104efef4a, 0x7}, 0x140010f7730)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/timeutils/stopwatch.go:111 +0x44
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0x0?)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_action.go:112 +0x78
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0x140014958a0)
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb0
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze in goroutine 102
        /Users/PYY388/go/pkg/mod/github.com/golangci/golangci-lint@v1.55.2/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x184 
WARN [runner] Can't run linter goanalysis_metalinter: goanalysis_metalinter: nilness: package "main" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled type *ir.SliceToArray 
INFO [runner] processing took 1.377µs with stages: max_same_issues: 458ns, skip_dirs: 167ns, max_from_linter: 83ns, severity-rules: 42ns, path_prefixer: 42ns, uniq_by_line: 42ns, diff: 42ns, filename_unadjuster: 42ns, nolint: 42ns, path_prettifier: 42ns, cgo: 42ns, skip_files: 42ns, autogenerated_exclude: 42ns, max_per_file_from_linter: 42ns, sort_results: 42ns, fixer: 42ns, exclude: 41ns, exclude-rules: 41ns, identifier_marker: 41ns, path_shortener: 0s, source_code: 0s 
INFO [runner] linters took 604.236458ms with stages: goanalysis_metalinter: 604.20725ms 
ERRO Running error: 1 error occurred:
        * can't run linter goanalysis_metalinter: goanalysis_metalinter: nilness: package "main" (isInitialPkg: true, needAnalyzeSource: true): internal error: unhandled type *ir.SliceToArray
 
INFO Memory: 9 samples, avg is 141.8MB, max is 227.7MB 
INFO Execution took 722.279208ms  
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
package main

import "fmt"

func main() {
	a, err := ToFixedBytes[[3]byte]([]byte{1, 2, 3})
	if err != nil {
		panic(err)
	}
	fmt.Println(a)
}

func ToFixedBytes[T [3]byte | [4]byte | [8]byte](b []byte) (T, error) {
	v := new(T)
	if len(b) != len(*v) {
		return *v, fmt.Errorf("invalid length: expected %d, got %d", len(*v), len(b))
	}
	return (T)(b), nil
}
```

</details>

### Validation

- [X] Yes, I've included all information above (version, config, etc.).

**Comment by ldez** [comment]
Hello,

this is a problem with [staticheck](https://github.com/dominikh/go-tools):
```console
$ staticcheck ./...                   
panic: internal error: unhandled type *ir.SliceToArray

goroutine 618 [running]:
honnef.co/go/tools/analysis/facts/nilness.impl.func1({0xa06ef0, 0xc0006bf570})
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:239 +0x774
honnef.co/go/tools/analysis/facts/nilness.impl.func1({0xa063e8, 0xc00081e980})
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:147 +0x8e7
honnef.co/go/tools/analysis/facts/nilness.impl(0xc000b78820, 0xc000a73680, 0xc00008ddb0)
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:246 +0x348
honnef.co/go/tools/analysis/facts/nilness.run(0xc000b78820)
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/analysis/facts/nilness/nilness.go:66 +0x110
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000ff1b30, {0xa04648?, 0xc0007b3360})
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/lintcmd/runner/runner.go:992 +0x6fd
honnef.co/go/tools/lintcmd/runner.genericHandle({0xa04648?, 0xc0007b3360}, {0xa04648?, 0xc000939220?}, 0x0?, 0xc00041e190, 0xc0003a2a90)
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/lintcmd/runner/runner.go:817 +0x129
created by honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers in goroutine 417
        /home/ldez/sources/go/pkg/mod/honnef.co/go/tools@v0.4.6/lintcmd/runner/runner.go:1061 +0x6b5
```

I tried with the [HEAD](https://github.com/dominikh/go-tools/commit/a367d57f6820a9e2458d1cc356efe3a979419721) of the branch master of [staticcheck](https://github.com/dominikh/go-tools), and the problem seems to be fixed.

When staticcheck will create a new release, our bot will update the dependecy and the problem will be fixed.
